### PR TITLE
[WIP][SPARK-28476][SQL] Support ALTER DATABASE SET LOCATION

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Support the syntax of ALTER (DATABASE|SCHEMA) database_name SET LOCATION path

Ref: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL

## How was this patch tested?

UT.

Please review https://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 1]
```diff
@@ -532,6 +532,22 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
...
+  /**
+   * Create an [[AlterDatabaseLocationCommand]] command.
+   *
+   * For example:
+   * {{{
+   *   ALTER (DATABASE|SCHEMA) database SET LOCATION path;
+   * }}}
+   */
+  override def visitSetDatabaseLocation(
+    ctx: SetDatabaseLocationContext): LogicalPlan = withOrigin(ctx) {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: add two more spaces

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -532,6 +532,22 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
...
+      AlterDatabaseLocationCommand(
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: remove two more spaces

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 3]
```diff
@@ -132,6 +132,27 @@ case class AlterDatabasePropertiesCommand(
...
+/**
+ * A command for users to set new location path for a database
+ * If the database does not exist, an error message will be issued to indicate the database
+ * does not exist.
+ * The syntax of using this command in SQL is:
+ * {{{
+ *    ALTER (DATABASE|SCHEMA) database_name SET LOCATION path
+ * }}}
+ */
+case class AlterDatabaseLocationCommand(databaseName: String, location: String)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Let us follow the name AlterTableSetLocationCommand. Rename it to AlterDatabaseSetLocationCommand

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLParserSuite.scala [hunk 4]
```diff
@@ -184,6 +184,15 @@ class DDLParserSuite extends AnalysisTest with SharedSQLContext {
       containsThesePhrases = Seq("key_without_value"))
   }
 
+  test("alter database set location") {
+    // ALTER (DATABASE|SCHEMA) database_name SET LOCATION
+    val sql1 = "ALTER DATABASE database_name SET LOCATION '/home/user/db'"
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLParserSuite.scala]: Add negative cases? For example, set an illegal uri string?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLParserSuite.scala]: Add it to DDLSuite?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 5]
```diff
@@ -532,6 +532,22 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
...
+   * For example:
+   * {{{
+   *   ALTER (DATABASE|SCHEMA) database SET LOCATION path;
+   * }}}
+   */
+  override def visitSetDatabaseLocation(
+      ctx: SetDatabaseLocationContext): LogicalPlan = withOrigin(ctx) {
+    AlterDatabaseLocationCommand(
+      ctx.db.getText,
+      visitLocationSpec(ctx.locationSpec())
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Nit: remove ()?

**Comment by gatorsmile** [comment]
cc @rednaxelafx

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLParserSuite.scala (hunk 4 above)
Comment (WeichenXu123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLParserSuite.scala]: I add a positive e2e test case.
What's the negative test case you want to add ?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLParserSuite.scala]: 1. A database that does not exist before issuing this command?
2. Setting an illegal path that could trigger an exception inside the function `stringToURI`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 6]
```diff
@@ -361,13 +361,16 @@ private[hive] class HiveClientImpl(
...
-      database.name,
+      dbName,
       new HiveDatabase(
-        database.name,
+        dbName,
         database.description,
-        CatalogUtils.URIToString(database.locationUri),
+        location,
         Option(database.properties).map(_.asJava).orNull))
+    client.getDatabase(dbName).setLocationUri(location)
```
Comment (WeichenXu123) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: CC @gatorsmile

**Comment by WeichenXu123** [comment]
Jenkins, retest this please.

**Comment by WeichenXu123** [comment]
Jenkins, retest this please.

**Comment by WeichenXu123** [comment]
@gatorsmile Any idea to address the hive metadata api "set db location" not working issue ?

**Comment by wangyum** [comment]
retest this please

**Pull request closed without merging**